File: Data Structures/Binary Search Tree/BST Views/BST_Views.java
import java.util.*;

class Node {
    int data;
    Node left, right;

    Node(int value) {
        data = value;
        left = right = null;
    }
}

class BST_Views {

    static class Pair {
        Node node;
        int hd;

        Pair(Node node, int hd) {
            this.node = node;
            this.hd = hd;
        }
    }

    public static List<Integer> topView(Node root) {
        Map<Integer, Integer> hdMap = new TreeMap<>();
        Queue<Pair> queue = new LinkedList<>();
        queue.add(new Pair(root, 0));
        while (!queue.isEmpty()) {
            Pair pair = queue.poll();
            if (!hdMap.containsKey(pair.hd)) {
                hdMap.put(pair.hd, pair.node.data);
            }
            if (pair.node.left != null) {
                queue.add(new Pair(pair.node.left, pair.hd - 1));
            }
            if (pair.node.right != null) {
                queue.add(new Pair(pair.node.right, pair.hd + 1));
            }
        }
        return new ArrayList<>(hdMap.values());
    }

    public static List<Integer> bottomView(Node root) {
        Map<Integer, Integer> hdMap = new TreeMap<>();
        Queue<Pair> queue = new LinkedList<>();
        queue.add(new Pair(root, 0));
        while (!queue.isEmpty()) {
            Pair pair = queue.poll();
            hdMap.put(pair.hd, pair.node.data);
            if (pair.node.left != null) {
                queue.add(new Pair(pair.node.left, pair.hd - 1));
            }
            if (pair.node.right != null) {
                queue.add(new Pair(pair.node.right, pair.hd + 1));
            }
        }
        return new ArrayList<>(hdMap.values());
    }

    public static List<Integer> leftView(Node root) {
        List<Integer> result = new ArrayList<>();
        Queue<Node> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            result.add(queue.peek().data);
            for (int i = 0, n = queue.size(); i < n; i++) {
                Node node = queue.poll();
                if (node.left != null) queue.add(node.left);
                if (node.right != null) queue.add(node.right);
            }
        }
        return result;
    }

    public static List<Integer> rightView(Node root) {
        List<Integer> result = new ArrayList<>();
        Queue<Node> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            result.add(queue.peek().data);
            for (int i = 0, n = queue.size(); i < n; i++) {
                Node node = queue.poll();
                if (node.left != null) queue.add(node.left);
                if (node.right != null) queue.add(node.right);
            }
        }
        return result;
    }

    public static Node buildTree() {
        try (Scanner sc = new Scanner(System.in)) {
            System.out.print("Enter the root node value: ");
            int value = sc.nextInt();
            Node root = new Node(value);

            Queue<Node> queue = new LinkedList<>();
            queue.add(root);

            while (!queue.isEmpty()) {
                Node current = queue.poll();

                System.out.print("Enter left child of " + current.data + " (-1 for no child): ");
                int leftValue = sc.nextInt();
                if (leftValue != -1) {
                    current.left = new Node(leftValue);
                    queue.add(current.left);
                }

                System.out.print("Enter right child of " + current.data + " (-1 for no child): ");
                int rightValue = sc.nextInt();
                if (rightValue != -1) {
                    current.right = new Node(rightValue);
                    queue.add(current.right);
                }
            }
            return root;
        }
    }

    public static void main(String[] args) {
        try (Scanner sc = new Scanner(System.in)) {
            System.out.println("Build the binary tree:");
            Node root = buildTree();

            while (true) {
                System.out.println("\nChoose the view to display:");
                System.out.println("1. Top View");
                System.out.println("2. Bottom View");
                System.out.println("3. Left View");
                System.out.println("4. Right View");
                System.out.println("5. Exit");

                System.out.print("Enter your choice: ");
                int choice = sc.nextInt();

                switch (choice) {
                    case 1 -> System.out.println("Top View: " + topView(root));
                    case 2 -> System.out.println("Bottom View: " + bottomView(root));
                    case 3 -> System.out.println("Left View: " + leftView(root));
                    case 4 -> System.out.println("Right View: " + rightView(root));
                    case 5 -> {
                        System.out.println("Exiting program.");
                        return;
                    }
                    default -> System.out.println("Invalid choice. Please try again.");
                }
            }
        }
    }
}

File: Data Structures/Binary Search Tree/BST Views/BST Views.cpp
#include <iostream>
#include <map>
#include <queue>
#include <vector>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;
    Node(int value) : data(value), left(nullptr), right(nullptr) {}
};

vector<int> topView(Node* root) {
    if (!root) return {};
    map<int, int> hdMap;
    queue<pair<Node*, int>> q;
    q.push({root, 0});
    while (!q.empty()) {
        auto [node, hd] = q.front();
        q.pop();
        if (!hdMap.count(hd)) hdMap[hd] = node->data;
        if (node->left) q.push({node->left, hd - 1});
        if (node->right) q.push({node->right, hd + 1});
    }
    vector<int> result;
    for (auto [_, val] : hdMap) result.push_back(val);
    return result;
}

vector<int> bottomView(Node* root) {
    if (!root) return {};
    map<int, int> hdMap;
    queue<pair<Node*, int>> q;
    q.push({root, 0});
    while (!q.empty()) {
        auto [node, hd] = q.front();
        q.pop();
        hdMap[hd] = node->data;
        if (node->left) q.push({node->left, hd - 1});
        if (node->right) q.push({node->right, hd + 1});
    }
    vector<int> result;
    for (auto [_, val] : hdMap) result.push_back(val);
    return result;
}

vector<int> leftView(Node* root) {
    if (!root) return {};
    vector<int> result;
    queue<Node*> q;
    q.push(root);
    while (!q.empty()) {
        result.push_back(q.front()->data);
        for (int i = 0, n = q.size(); i < n; ++i) {
            Node* node = q.front();
            q.pop();
            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
        }
    }
    return result;
}

vector<int> rightView(Node* root) {
    if (!root) return {};
    vector<int> result;
    queue<Node*> q;
    q.push(root);
    while (!q.empty()) {
        result.push_back(q.back()->data);
        for (int i = 0, n = q.size(); i < n; ++i) {
            Node* node = q.front();
            q.pop();
            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
        }
    }
    return result;
}

Node* buildTree() {
    int value;
    cout << "Enter the root node value: ";
    cin >> value;
    Node* root = new Node(value);

    queue<Node*> q;
    q.push(root);

    while (!q.empty()) {
        Node* current = q.front();
        q.pop();

        cout << "Enter left child of " << current->data << " (-1 for no child): ";
        cin >> value;
        if (value != -1) {
            current->left = new Node(value);
            q.push(current->left);
        }

        cout << "Enter right child of " << current->data << " (-1 for no child): ";
        cin >> value;
        if (value != -1) {
            current->right = new Node(value);
            q.push(current->right);
        }
    }

    return root;
}

int main() {
    Node* root = buildTree();

    int choice;
    do {
        cout << "\nChoose the view to display:\n";
        cout << "1. Top View\n";
        cout << "2. Bottom View\n";
        cout << "3. Left View\n";
        cout << "4. Right View\n";
        cout << "5. Exit\n";
        cout << "Enter your choice: ";
        cin >> choice;

        vector<int> result;

        switch (choice) {
            case 1:
                result = topView(root);
                cout << "Top View: ";
                for (int val : result) cout << val << " ";
                cout << endl;
                break;
            case 2:
                result = bottomView(root);
                cout << "Bottom View: ";
                for (int val : result) cout << val << " ";
                cout << endl;
                break;
            case 3:
                result = leftView(root);
                cout << "Left View: ";
                for (int val : result) cout << val << " ";
                cout << endl;
                break;
            case 4:
                result = rightView(root);
                cout << "Right View: ";
                for (int val : result) cout << val << " ";
                cout << endl;
                break;
            case 5:
                cout << "Exiting the program." << endl;
                break;
            default:
                cout << "Invalid choice. Please try again." << endl;
        }
    } while (choice != 5);

    return 0;
}
